libsolidity/grammar.txt

```diff
@@ -1,48 +1,96 @@
-ContractDefinition = ( 'contract' | 'library' ) Identifier
+ContractDefinition = ('import' Imports ';')*
+                     ( 'contract' | 'library' ) Identifier
                      ( 'is' InheritanceSpecifier (',' InheritanceSpecifier )* )?
                      '{' ContractPart* '}'
-ContractPart = StateVariableDeclaration | StructDefinition | ModifierDefinition | FunctionDefinition | EnumDefinition
+
+Imports = StringLiteral ('as' Identifier)?
+        | ('*' | Identifier) ('as' Identifier)? 'from' StringLiteral
+        | '{' Identifier ('as' Identifier)? ( ',' Identifier ('as' Identifier)? )* '}' 'from' StringLiteral
+
+ContractPart = StateVariableDecOrDef | UsingDeclaration
+             | StructDefinition | ModifierDefinition | FunctionDefinition | EventDefinition | EnumDefinition
 
 InheritanceSpecifier = Identifier ( '(' Expression ( ',' Expression )* ')' )?
+
+StateVariableDecOrDef = TypeName ( 'public' | 'internal' | 'private' )? Identifier ('=' Expression)? ';'
+UsingDeclaration = 'using' Identifier 'for' TypeName ';'
 StructDefinition = 'struct' Identifier '{'
-                     ( VariableDeclaration (';' VariableDeclaration)* )? '}
-StateVariableDeclaration = TypeName ( 'public' | 'inheritable' | 'private' )? Identifier ';'
+                     ( VariableDeclaration ';' (VariableDeclaration ';')* )? '}'
 ModifierDefinition = 'modifier' Identifier ParameterList? Block
 FunctionDefinition = 'function' Identifier ParameterList
-					 ( Identifier | 'constant' | 'external' | 'public' | 'inheritable' | 'private' )*
-                     ( 'returns' ParameterList )? Block
+                     ( FunctionCall | Identifier | 'constant' | 'external' | 'public' | 'internal' | 'private' )*
+                     ( 'returns' (ParameterList | TypeParameterList) )? Block
+EventDefinition = 'event' Identifier (ParameterList | TypeParameterList | IndexedTypeParameterList) 'anonymous'? ';'
 
 EnumValue = Identifier
-EnumDefinition = 'enum' '{' EnumValue (',' EnumValue)* '}'
+EnumDefinition = 'enum' Identifier '{' EnumValue? (',' EnumValue)* '}'
+
+IndexedTypeParameterList = '(' ( TypeName 'indexed'? Identifier? (',' TypeName 'indexed'? Identifier?)* )? ')'
+TypeParameterList = '(' ( TypeName (',' TypeName)* )? ')'
 ParameterList = '(' ( VariableDeclaration (',' VariableDeclaration)* )? ')'
 // semantic restriction: mappings and structs (recursively) containing mappings
 // are not allowed in argument lists
-VariableDeclaration = TypeName Identifier
-TypeName = ElementaryTypeName | Identifier | Mapping | ArrayTypeName
+VariableDeclaration = TypeName Identifier | 'var' assignment
```

Comment:
`var` is only allowed for local variables, but I'm not sure if we should enforce that on the grammar level (the parser does) and that might also be lifted in the future.